### ISSUE TITLE
ability to validate UTF-8 encoding

### ISSUE THREAD

**Issue by andrewrk** [ashtuchkin/iconv-lite#83]
In iconv:

```
> Iconv = require('iconv').Iconv
> i = new Iconv('UTF-8', 'UTF-8')
> i.convert(new Buffer([128]))
Error: Illegal character sequence.
```

In iconv-lite:

```
> iconv = require('iconv-lite')
> iconv.decode(new Buffer([128]), 'UTF-8')
'�'
```

With iconv, you know whether the decoding succeeded. With iconv-lite there is no way to know if the decoding succeeded.

**Comment by bpasero** [comment]
@ashtuchkin I was thinking of a similar request today. Is there any way to know (after a call to encode) that the output is bogus? I think iconv is more agressive on throwing an error if the encoding choice does not make sense for the input. See https://github.com/bnoordhuis/node-iconv#dealing-with-untranslatable-characters

**Comment by ashtuchkin** [comment]
Like I said above, at the moment, the only way to do that is to check for the presence of character � in result. The thing is, I don't control UTF-8 decoding - it's done by Node.js. I could reimplement it in javascript, but I think it'd be a lot slower.

**Comment by bpasero** [comment]
@ashtuchkin got it. But in the docs you say "Untranslatable characters are set to � or ?", so it might also be "?" which I cannot really check for because its a valid character.

**Comment by ashtuchkin** [comment]
"�" is used for decoding, "?" for encoding (as many encodings cannot
represent �).

--
Alexander Shtuchkin

**Comment by bpasero** [comment]
Right, but "?" is ambiguos because it might also be a character used in the original text. I think raising some flag or error would make more sense to catch this case.

**Comment by ashtuchkin** [comment]
1. Are you talking about encoding or decoding? I thought decoding, right?
2. Uft8 is still outside my control.

**Comment by bpasero** [comment]
In our case, we use encode() to convert an existing file to another encoding. E.g. we have a source UTF-16 file that we want to save as DOS encoding. DOS can not present all encodings (Chinese et. al), so I would like to show an error to the user. Similar to how in Sublime Text you will not be able to save a file in an encoding that cannot represent all characters you have before encoding.

**Comment by ashtuchkin** [comment]
Yeah, that's a completely separate issue from "ability to validate UTF-8
encoding" :)

I have a callback-like mechanism in mind (see #53), where you can either
throw exception or define replacements for characters that cannot be
represented. Still haven't got a moment to implement it though.​

--
Alexander Shtuchkin

**Comment by bpasero** [comment]
Makes sense!

**Comment by FazilMuhammed** [comment]
i cannot encode some characters like " or ' what is the reason any one help me ? i need Arabic text also need to print.

**Comment by yosion-p** [comment]
As I slowly scrolled through the page, I thought about how this problem could be solved.
To my surprise, our idea is consistent.
> I have a callback-like mechanism in mind (see #53), where you can either
I think many issues are the same as 53, it's about CALLBACK.
I will try to solve an issue with CB on weekends.
